openstack/standalone-cinder/provisioner.go

```diff
@@ -0,0 +1,224 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+
+	"github.com/golang/glog"
+	"github.com/kubernetes-incubator/external-storage/lib/controller"
+	"github.com/gophercloud/gophercloud"
+
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/api/core/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const (
+	provisionerName  = "openstack.org/standalone-cinder"
+	provisionerIDAnn = "standaloneCinderProvisionerIdentity"
+	cinderVolumeId = "cinderVolumeId"
+)
+
+type cinderProvisioner struct {
+	// Openstack cinder client
+	volumeService *gophercloud.ServiceClient
+
+	// Kubernetes Client. Use to create secret
+	client kubernetes.Interface
+	// Identity of this cinderProvisioner, generated. Used to identify "this"
+	// provisioner's PVs.
+	identity string
+}
+
+func newCinderProvisioner(client kubernetes.Interface, id, configFilePath string) (controller.Provisioner, error) {
+	volumeService, err := getVolumeService(configFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get volume service: %v", err)
+	}
+
+	return &cinderProvisioner{
+		volumeService: volumeService,
+		client:   client,
+		identity: id,
+	}, nil
+}
+
+var _ controller.Provisioner = &cinderProvisioner{}
+
+
+type provisionCtx struct {
+	p *cinderProvisioner
+	options controller.VolumeOptions
+	connection volumeConnection
+}
+
+
+type deleteCtx struct {
+	p *cinderProvisioner
+	pv *v1.PersistentVolume
+}
+
+
+// Provision creates a storage asset and returns a PV object representing it.
+func (p *cinderProvisioner) Provision(options controller.VolumeOptions) (*v1.PersistentVolume, error) {
+	if options.PVC.Spec.Selector != nil {
+		return nil, fmt.Errorf("claim Selector is not supported")
+	}
+
+	volumeId, err := createCinderVolume(p, options)
+	if err != nil {
+		glog.Errorf("Failed to create volume")
+		return nil, err
+	}
+
+	connection, err := connectCinderVolume(p, volumeId)
+	if err != nil {
+		// TODO: Create placeholder PV?
+		glog.Errorf("Failed to connect volume: %v", err)
+		return nil, err
+	}
+
```

Comment:
reserve the volume

Reply:
I would like to add TODOs in the code for this.  We should also call terminate_connection to tear down iscsi lun mappings on the storage server for example.

Reply:
sgtm